### ISSUE TITLE
Why are there no CPU-based 3D algorithms?

### ISSUE THREAD

**Issue by pierrethibault** [astra-toolbox/astra-toolbox#158]
I was just wondering. There has to be users with limited or no access to GPUs.

**Comment by AnderBiguri** [comment]
@pierrethibault  because it would take days at best, but easily weeks to run in a normal computer.

**Comment by tokkot** [comment]
I agree that CPU algorithms in 3D would be useful, and, as far as I know, there is no good implementation out there (as flexible as astra).

@AnderBiguri I don't think it would be as slow as you imagine. For, say, a 100x100x100 image, I've seen STIR project 1 million rays in minutes. I think this would be typical for a decent implementation of SIddon's algorithm.

**Comment by AnderBiguri** [comment]
@tokkot 100x100x100 is tiny even for a toy example. ASTRA backprojects/projects 512^3 images with 512^2*360 projections in less than 5 seconds. 

This is a "normal" medical scan that would take hours, possibly days to project using the CPUspeeds you suggest.

**Comment by tokkot** [comment]
@AnderBiguri Let's agree that medical reconstructions are in the range of some hundreds of pixels cubed (PET scans seem to get close to the 100x100x100 range, so I would not consider it tiny). I also agree that the GPU implementation is likely to be much faster than a CPU one.

My point is only that there are some researchers who do CPU-based reconstructions in this range of sizes. In my experience, these reconstructions are in the hours/days range, and we are willing to wait. We can also do algorithm development on toy-sized problems. If CPU projectors were added to the astra project, I would use them.

**Comment by AnderBiguri** [comment]
@tokkot Yes I realized STIR is more for PET and SPECT, thus is not that tiny (thus my sneaky addition "in CT" there :) ).

You are right, research timeline is different that "real life" timeline (I should know, I do CT reconstruction research!) . You are right that the CPU code would help people, I was just arguing why its probably not there and why it will probably not be (same reason why I wont add it to my own rec toolbox!). The question was "Why?" and perhaps I took it a bit literally. 

Anyway, there are few CPU based codes around the web, some googling may show them (I dont want to advertise them here).
